### PR TITLE
Node and Pod Metrics

### PR DESCRIPTION
**Pull request by r-dilip**

### CODE REVIEW

File: installer/conf/kube.conf [hunk 1]
```diff
@@ -119,8 +125,8 @@
       max_retry_wait 9m
      </match>
 
-     <match oms.api.ContainerNodeInventory**>
-      type out_oms_api
+     <match oms.containerinsights.ContainerNodeInventory**>
```
Comment (vishiy) [inline installer/conf/kube.conf]: whay are you changing this ? Th was Rashmi's change that went in few weeks back into ci_feature. Can you merge from ci_feature into your branch and then create a PR ?

File: installer/datafiles/base_container.data [hunk 2]
```diff
@@ -177,6 +178,10 @@ touch /var/opt/microsoft/docker-cimprov/log/filter_cadvisor2mdm.log
 chmod 666 /var/opt/microsoft/docker-cimprov/log/filter_cadvisor2mdm.log
 chown omsagent:omiusers /var/opt/microsoft/docker-cimprov/log/filter_cadvisor2mdm.log
 
+touch /var/opt/microsoft/docker-cimprov/log/filter_inventory2mdm.log
```
Comment (vishiy) [inline installer/datafiles/base_container.data]: why do we need this for this plugin ? Also why 'write' perms ? (666)

File: installer/conf/kube.conf (hunk 1 above)
Comment (r-dilip) [inline installer/conf/kube.conf]: I still need to merge ci_feature into feature/customMetrics. That is why this change is seen. I merged from ci_feature into my topic branch

File: source/code/plugin/filter_cadvisor2mdm.rb [hunk 3]
```diff
@@ -43,6 +43,12 @@ class CAdvisor2MdmFilter < Filter
                     } 
                 } 
             }'
+        
+        @@metric_name_metric_percentage_name_hash = {
```
Comment (vishiy) [inline source/code/plugin/filter_cadvisor2mdm.rb]: do this need to be static ?

File: source/code/plugin/filter_inventory2mdm.rb [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                @log.debug "Returning false for check_custom_metrics_availability"
+                return false
+            end
+        end
+
+		def shutdown
+			super
+        end
+        
+        # KUBE_NODE_INVENTORY
```
Comment (vishiy) [inline source/code/plugin/filter_inventory2mdm.rb]: Do we need this payload ? Also it has subscription info (which will be open to all in github)

File: source/code/plugin/filter_inventory2mdm.rb [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+			end
+		end
+
+        def start
+            super
+            @process_incoming_stream = check_custom_metrics_availability
+            @log.debug "After check_custom_metrics_availability process_incoming_stream #{@process_incoming_stream}"
+        end
+
+        def check_custom_metrics_availability
```
Comment (vishiy) [inline source/code/plugin/filter_inventory2mdm.rb]: This function is in multiple places now. Good reason to put it as a library & ref it..

File: source/code/plugin/filter_inventory2mdm.rb [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    require_relative 'oms_common'
+
+	class Inventory2MdmFilter < Filter
+		Fluent::Plugin.register_filter('filter_inventory2mdm', self)
+		
+		config_param :enable_log, :integer, :default => 0
+        config_param :log_path, :string, :default => '/var/opt/microsoft/docker-cimprov/log/filter_inventory2mdm.log'
+        config_param :custom_metrics_azure_regions, :string
+        
+        @@node_count_metric_name = 'nodesCount'
```
Comment (vishiy) [inline source/code/plugin/filter_inventory2mdm.rb]: I am thinking if we should name these as 'numPods' & 'numNodes'

File: source/code/plugin/filter_inventory2mdm.rb [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+                } 
+            }'
+
+        @@pod_inventory_custom_metrics_template = '
+            { 
+                "time": "%{timestamp}", 
+                "data": { 
+                    "baseData": { 
+                        "metric": "%{metricName}", 
+                        "namespace": "insights.container/nodes",
```
Comment (vishiy) [inline source/code/plugin/filter_inventory2mdm.rb]: can this go to 'pods' namespace instead of 'nodes' ?

File: source/code/plugin/filter_cadvisor2mdm.rb (hunk 3 above)
Comment (r-dilip) [reply source/code/plugin/filter_cadvisor2mdm.rb]: its a static lookup that doesnt change.

File: installer/conf/kube.conf (hunk 1 above)
Comment (r-dilip) [reply installer/conf/kube.conf]: github PRs are static. So it doesnt get updated even after I merged and pushed. So ignore this change.

File: source/code/plugin/filter_inventory2mdm.rb (hunk 6 above)
Comment (r-dilip) [reply source/code/plugin/filter_inventory2mdm.rb]: Why? I think nodeCount and podCount is just fine

File: source/code/plugin/filter_inventory2mdm.rb (hunk 7 above)
Comment (r-dilip) [reply source/code/plugin/filter_inventory2mdm.rb]: that will create too many namespaces. I think it makes sense to have one single namespace for container insights. This will also cause discoverability issues. We can't expect customers to go check every namespace to see what metrics each one has. Just my thoughts

File: source/code/plugin/in_kube_nodes.rb [hunk 8]
```diff
@@ -109,7 +110,12 @@ def enumerate
                     dockerVersion.slice! "docker://"
                     containerNodeInventoryRecord['DockerVersion'] = dockerVersion
                     # ContainerNodeInventory data for docker version and operating system.
-                    containerNodeInventoryEventStream.add(emitTime, containerNodeInventoryRecord) if containerNodeInventoryRecord
+                    containerNodeInventoryWrapper = {
```
Comment (r-dilip) [inline source/code/plugin/in_kube_nodes.rb]: Ignore this change. Since the PR was created before the target branch was updated, this shows up.

File: source/code/plugin/filter_inventory2mdm.rb (hunk 6 above)
Comment (vishiy) [reply source/code/plugin/filter_inventory2mdm.rb]: That seems to be the convention for 'count' metrics. But i am okay to leave it as is until somebody tells so :)

File: source/code/plugin/filter_inventory2mdm.rb (hunk 7 above)
Comment (vishiy) [reply source/code/plugin/filter_inventory2mdm.rb]: #1 ) There are not much restrictions on the num namespaces
#2) So you expect customers to look for podcount under /nodes (and will make it easier to discover than putting them under 'pods' ? In the futuire when we add pod cpu, we will add it under nodes? then how about node cpu ? If i am looking for charting anything podlevel its easy to understand to go to pods namespace.
3) either flatten everything and put under default namespace or catagorize it properly.
